PurchasesCoreSwift/Purchasing/ReceiptFetcher.swift

```diff
@@ -0,0 +1,95 @@
+@objc(RCReceiptFetcher) public class ReceiptFetcher: NSObject {
+    private let requestFetcher: StoreKitRequestFetcher
+    private let receiptBundle: Bundle
+
+    @objc public convenience init(requestFetcher: StoreKitRequestFetcher) {
+        self.init(requestFetcher: requestFetcher, bundle: .main)
+    }
+
+    init(requestFetcher: StoreKitRequestFetcher, bundle: Bundle) {
+        self.requestFetcher = requestFetcher
+        self.receiptBundle = bundle
+    }
+
+    @objc public func receiptData(refreshPolicy: ReceiptRefreshPolicy,
+                                  completion: @escaping ((Data?) -> Void)) {
+        if refreshPolicy == .always {
+            Logger.debug(String(format: Strings.receipt.force_refreshing_receipt))
+            self.refreshReceipt(completion)
+            return
+        }
+
+        let receiptData = receiptData()
+        let isReceiptEmpty = receiptData?.isEmpty ?? true
+
+        if isReceiptEmpty && refreshPolicy == .onlyIfEmpty {
+            Logger.debug(Strings.receipt.refreshing_empty_receipt)
+            self.refreshReceipt(completion)
+        } else {
+            completion(receiptData)
+        }
+    }
+
+}
+
+private extension ReceiptFetcher {
+
+    func receiptData() -> Data? {
+        guard var receiptURL: URL = self.receiptBundle.appStoreReceiptURL else {
+            Logger.debug(Strings.receipt.no_sandbox_receipt_restore)
+            return nil
+        }
+
+        #if os(watchOS)
+        // as of watchOS 6.2.8, there's a bug where the receipt is stored in the sandbox receipt location,
+        // but the appStoreReceiptURL method returns the URL for the production receipt.
+        // This code replaces "sandboxReceipt" with "receipt" as the last component of the receiptURL so that we get the
+        // correct receipt.
+        // This has been filed as radar FB7699277. More info in https://github.com/RevenueCat/purchases-ios/issues/207.
+
+        let minimumOSVersionWithoutBug: OperatingSystemVersion = OperatingSystemVersion(majorVersion: 7, minorVersion: 0, patchVersion: 0)
+        let isBelowMinimumOSVersionWithoutBug: Bool = ProcessInfo.processInfo.isOperatingSystemAtLeast(minimumOSVersionWithoutBug)
+
+        if isBelowMinimumOSVersionWithoutBug && SystemInfo.isSandbox {
+            let receiptURLFolder: URL = receiptURL.deletingLastPathComponent()
+            let productionReceiptURL: URL = receiptURLFolder.appendingPathComponent("receipt")
+            receiptURL = productionReceiptURL
+        }
+        #endif
+
+        guard let data: Data = try? Data(contentsOf: receiptURL) else {
+            Logger.debug(Strings.receipt.unable_to_load_receipt)
+            return nil
+        }
+
+        Logger.debug(String(format: Strings.receipt.loaded_receipt, receiptURL as CVarArg))
+
+        return data
+    }
+
+    func refreshReceipt(_ completion: @escaping ((Data) -> Void)) {
+        requestFetcher.fetchReceiptData {
+            let maybeData = self.receiptData()
```

Comment:
Well, I don't know if I should review your code 😅 But I was checking this PR, and I found this line, and I think this could be a potential retain cycle because the closure is retaining `self`, doesn't it? 🤔

Reply:
Feel free to review any of PRs @Juanpe 😃 

If it caused a retain cycle, it would be temporary while the request is in flight. Though, I can't see anywhere that these blocks are being retained outside of this function, so *I think* this is safe. That being said, we could capture `self` weakly to be sure, but I don't think that's totally necessary.

Reply:
Yep, you're right 🙂 In this case, @aboedo forget this comment

![](https://media.giphy.com/media/R7m04yMaGWVeE/giphy.gif)

Reply:
reviews are always welcome @juanpe! 

And yes, the retain here is only until `fetchReceiptData` calls completion, so the question there is 
"if this object is no longer referenced by the time the completion comes back, should we move forward?"

In practice, within our SDK, this would only happen if you call `configure` again and get a new instance. Which... you shouldn't do. And we might try to make that impossible to do in the next major. 

If you did, however, the objects from the old instance will be dealloc'ed once all their stuff is no longer referenced, so I think for stuff like this, what would happen is that the method finishes executing and releases the instance, and in turn whoever was waiting for the receipt data would get it, and probably post it to the backend, which is fine, as the data should still be consistent. 

All of that said, do you think it's worth having a `[self]` in the capture list to be more explicit about what the plan is? 

Sidenote: FWIW, after a few years of working with RxSwift I got used to getting alarmed whenever a strong self is captured in a block, but for non-reactive programming it's often okay.

Reply:
Thanks for the clarification 👌🏼

Re `[self]`: I think it's not needed, after the explanation I see your point. 
[Totally minor] I would just make sure that we use the same way every time `self` is captured in a block, for example both `HTTPClient` and `ProductsManager` classes have some `[self]` inside. Otherwise, it could be confused for future contributors, devs, etc. I mean, they could think... "Why here and why not ... maybe there is a reason for that."

PS: I'm working with RxSwift on my current project and maybe I'm getting used to being alarmed every time a strong me is captured in a block 😅 😂

Reply:
I agree, I think consistency here is important. 
I'm gonna add a note to add it to the [swift style guide](https://github.com/RevenueCat/purchases-ios/blob/swift_migration/SwiftStyleGuide.swift)

Reply:
Thanks! 🤙

Reply:
https://github.com/RevenueCat/purchases-ios/pull/714